File: app/src/test/java/com/parknav/simulation/PathSimulator.kt
package com.parknav.simulation

import com.parknav.data.Position
import com.parknav.sensor.SensorData
import kotlin.math.*

/**
 * Simulated path with ground truth for validation.
 */
data class SimulatedPath(
    val sensorData: List<SensorData>,
    val groundTruth: List<Position>,
    val totalDistance: Float,
    val description: String
)

/**
 * Generates simulated IMU sensor data for known paths.
 *
 * Coordinate system matches PDREngine:
 *   X = east (heading=π/2), Z = south (heading=π), north = heading 0
 *   posX += stepLength * sin(heading)
 *   posZ += stepLength * cos(heading)
 *
 * Magnetometer encoding for heading θ (radians, 0=north clockwise):
 *   magX = 20 * sin(θ)   (east component)
 *   magY = 20 * cos(θ)   (north component)
 *   magZ = -40            (downward, typical)
 *
 * This makes atan2(magY, magX) recover the heading in the way
 * TestableOrientationEstimator expects.
 */
object PathSimulator {

    private const val SAMPLE_INTERVAL_NS = 20_000_000L // 20ms = 50Hz
    private const val GRAVITY = 9.81f
    private const val WALK_FREQ = 2.0    // 2 Hz step frequency
    private const val WALK_AMP = 4.0     // m/s² amplitude on top of gravity
    private const val MAG_HORIZONTAL = 20f
    private const val MAG_VERTICAL = -40f

    // ── public API ────────────────────────────────────────────────

    fun simulateStraightWalk(
        distance: Float,
        heading: Float,      // radians, 0 = north
        stepLength: Float = 0.7f
    ): SimulatedPath {
        val steps = (distance / stepLength).roundToInt()
        val stepsPerSec = WALK_FREQ
        val walkDurationMs = (steps / stepsPerSec * 1000).toLong()
        val totalSamples = (walkDurationMs / 20).toInt()

        val sensorData = mutableListOf<SensorData>()
        val groundTruth = mutableListOf(Position(0f, 0f, 0f)) // start

        val dx = stepLength * sin(heading)
        val dz = stepLength * cos(heading)

        var currentX = 0f
        var currentZ = 0f
        var stepIndex = 0

        for (i in 0 until totalSamples) {
            val t = i * 0.02 // seconds
            val ts = i * SAMPLE_INTERVAL_NS

            // Walking waveform – sinusoidal on the Y-axis (vertical in portrait hold)
            val walkComponent = WALK_AMP * sin(2 * PI * WALK_FREQ * t)
            val accY = (GRAVITY + walkComponent).toFloat()

            sensorData += SensorData(
                accX = 0f, accY = accY, accZ = 0f,
                gyroX = 0f, gyroY = 0f, gyroZ = 0f,
                magX = MAG_HORIZONTAL * sin(heading),
                magY = MAG_HORIZONTAL * cos(heading),
                magZ = MAG_VERTICAL,
                timestamp = ts
            )

            // Ground-truth position: advance one step at each peak
            // Peak occurs when sin(...) = 1 → t = (0.25 + n) / freq
            val cyclePos = (WALK_FREQ * t) % 1.0
            if (cyclePos in 0.24..0.26 && stepIndex < steps) {
                stepIndex++
                currentX += dx
                currentZ += dz
                groundTruth += Position(currentX, 0f, currentZ)
            }
        }

        // Ensure we have the final position
        if (stepIndex < steps) {
            currentX = steps * dx
            currentZ = steps * dz
            groundTruth += Position(currentX, 0f, currentZ)
        }

        return SimulatedPath(
            sensorData = sensorData,
            groundTruth = groundTruth,
            totalDistance = distance,
            description = "Straight walk %.0fm heading %.0f°".format(distance, Math.toDegrees(heading.toDouble()))
        )
    }

    fun simulateTurn(
        fromHeading: Float,
        toHeading: Float,
        steps: Int = 3
    ): SimulatedPath {
        // Turn over `steps` walking steps, smoothly rotating heading
        val stepsPerSec = WALK_FREQ
        val durationMs = (steps / stepsPerSec * 1000).toLong()
        val totalSamples = (durationMs / 20).toInt()

        val sensorData = mutableListOf<SensorData>()
        val groundTruth = mutableListOf(Position(0f, 0f, 0f))

        // Shortest angular difference
        var delta = toHeading - fromHeading
        while (delta > PI) delta -= (2 * PI).toFloat()
        while (delta < -PI) delta += (2 * PI).toFloat()

        val angularRate = delta / (durationMs / 1000f) // rad/s

        var currentX = 0f
        var currentZ = 0f
        var stepIndex = 0
        val stepLength = 0.7f

        for (i in 0 until totalSamples) {
            val t = i * 0.02
            val ts = i * SAMPLE_INTERVAL_NS
            val progress = (t / (durationMs / 1000.0)).coerceIn(0.0, 1.0).toFloat()
            val currentHeading = fromHeading + delta * progress

            val walkComponent = WALK_AMP * sin(2 * PI * WALK_FREQ * t)
            val accY = (GRAVITY + walkComponent).toFloat()

            sensorData += SensorData(
                accX = 0f, accY = accY, accZ = 0f,
                gyroX = 0f, gyroY = 0f, gyroZ = angularRate,
                magX = MAG_HORIZONTAL * sin(currentHeading),
                magY = MAG_HORIZONTAL * cos(currentHeading),
                magZ = MAG_VERTICAL,
                timestamp = ts
            )

            val cyclePos = (WALK_FREQ * t) % 1.0
            if (cyclePos in 0.24..0.26 && stepIndex < steps) {
                stepIndex++
                currentX += stepLength * sin(currentHeading)
                currentZ += stepLength * cos(currentHeading)
                groundTruth += Position(currentX, 0f, currentZ)
            }
        }

        return SimulatedPath(
            sensorData = sensorData,
            groundTruth = groundTruth,
            totalDistance = steps * stepLength,
            description = "Turn from %.0f° to %.0f° in %d steps".format(
                Math.toDegrees(fromHeading.toDouble()),
                Math.toDegrees(toHeading.toDouble()),
                steps
            )
        )
    }

    fun simulateLShape(legLength1: Float, legLength2: Float): SimulatedPath {
        val heading1 = 0f              // north
        val heading2 = (PI / 2).toFloat() // east (90° right turn)

        val leg1 = simulateStraightWalk(legLength1, heading1)
        val turn = simulateTurn(heading1, heading2, steps = 3)
        val leg2 = simulateStraightWalk(legLength2, heading2)

        return combinePaths(
            listOf(leg1, turn, leg2),
            "L-shape: %.0fm north + %.0fm east".format(legLength1, legLength2),
            legLength1 + legLength2
        )
    }

    fun simulateRectangle(width: Float, height: Float): SimulatedPath {
        val headings = listOf(0f, (PI / 2).toFloat(), PI.toFloat(), (3 * PI / 2).toFloat())
        val lengths = listOf(height, width, height, width)

        val segments = mutableListOf<SimulatedPath>()
        for (i in 0 until 4) {
            segments += simulateStraightWalk(lengths[i], headings[i])
            if (i < 3) {
                segments += simulateTurn(headings[i], headings[i + 1], steps = 2)
            } else {
                segments += simulateTurn(headings[3], headings[0], steps = 2)
            }
        }

        return combinePaths(
            segments,
            "Rectangle %.0fm × %.0fm".format(width, height),
            2 * (width + height)
        )
    }

    fun simulateStationary(durationMs: Long): SimulatedPath {
        val totalSamples = (durationMs / 20).toInt()
        val sensorData = (0 until totalSamples).map { i ->
            SensorData(
                accX = 0f, accY = GRAVITY, accZ = 0f,
                gyroX = 0f, gyroY = 0f, gyroZ = 0f,
                magX = 0f, magY = MAG_HORIZONTAL, magZ = MAG_VERTICAL,
                timestamp = i * SAMPLE_INTERVAL_NS
            )
        }

        return SimulatedPath(
            sensorData = sensorData,
            groundTruth = listOf(Position(0f, 0f, 0f)),
            totalDistance = 0f,
            description = "Stationary %dms".format(durationMs)
        )
    }

    // ── helpers ───────────────────────────────────────────────────

    private fun combinePaths(
        paths: List<SimulatedPath>,
        description: String,
        totalDistance: Float
    ): SimulatedPath {
        val allSensor = mutableListOf<SensorData>()
        val allTruth = mutableListOf<Position>()

        var timeOffset = 0L
        var posOffsetX = 0f
        var posOffsetZ = 0f

        for (path in paths) {
            // Append sensor data with shifted timestamps
            for (d in path.sensorData) {
                allSensor += d.copy(timestamp = d.timestamp + timeOffset)
            }
            val maxTs = path.sensorData.maxOfOrNull { it.timestamp } ?: 0L
            timeOffset += maxTs + SAMPLE_INTERVAL_NS

            // Append ground truth with position offset (skip first = origin of segment)
            val truthToAdd = if (allTruth.isEmpty()) path.groundTruth else path.groundTruth.drop(1)
            for (p in truthToAdd) {
                allTruth += Position(p.x + posOffsetX, 0f, p.z + posOffsetZ)
            }

            // Update offset from last truth point of this segment
            if (path.groundTruth.isNotEmpty()) {
                val last = path.groundTruth.last()
                posOffsetX += last.x
                posOffsetZ += last.z
            }
        }

        return SimulatedPath(
            sensorData = allSensor,
            groundTruth = allTruth,
            totalDistance = totalDistance,
            description = description
        )
    }

    private fun Float.roundToInt(): Int = kotlin.math.round(this).toInt()
}

File: simulator/path_simulator.py
"""
Path simulator — generates realistic IMU sensor streams for known paths.
Sampling rate: 50 Hz (20 ms).  Walking frequency: 2 Hz.
"""

import math
import numpy as np
from typing import List, Tuple
from pdr_engine import SensorData

SAMPLE_RATE = 50        # Hz
DT_NS = int(1e9 / SAMPLE_RATE)
WALK_FREQ = 2.0         # Hz (step frequency)
WALK_AMP = 1.92          # m/s² amplitude; Weinberg K=0.5 → step ≈ 0.7m
GRAVITY = 9.81
MAG_STRENGTH = 20.0

# Noise std devs (defaults)
ACC_NOISE = 0.3
GYRO_NOISE = 0.01
MAG_NOISE = 1.0


def _walk_samples(duration_s: float, heading: float, t0_ns: int = 0,
                  noise_scale: float = 1.0) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    """Generate sensor data for walking at constant heading for *duration_s* seconds."""
    n = int(duration_s * SAMPLE_RATE)
    samples: List[SensorData] = []
    # Ground-truth waypoints: one per step
    gt_points: List[Tuple[float, float]] = []

    for i in range(n):
        t_s = i / SAMPLE_RATE
        ts = t0_ns + i * DT_NS

        # Accelerometer: gravity + walking oscillation along vertical
        osc = WALK_AMP * math.sin(2 * math.pi * WALK_FREQ * t_s)
        acc_mag = GRAVITY + osc
        # Distribute on axes assuming phone roughly upright
        acc_x = acc_mag * math.sin(heading) * 0.1 + np.random.normal(0, ACC_NOISE * noise_scale)
        acc_y = acc_mag * 0.99 + np.random.normal(0, ACC_NOISE * noise_scale)
        acc_z = acc_mag * math.cos(heading) * 0.1 + np.random.normal(0, ACC_NOISE * noise_scale)

        # Magnetometer
        mag_x = MAG_STRENGTH * math.sin(heading) + np.random.normal(0, MAG_NOISE * noise_scale)
        mag_y = MAG_STRENGTH * math.cos(heading) + np.random.normal(0, MAG_NOISE * noise_scale)
        mag_z = np.random.normal(0, MAG_NOISE * noise_scale)

        # Gyroscope — straight walk, nearly zero yaw rate
        gyro_z = np.random.normal(0, GYRO_NOISE * noise_scale)

        samples.append(SensorData(
            timestamp=ts,
            accX=acc_x, accY=acc_y, accZ=acc_z,
            gyroX=0.0, gyroY=0.0, gyroZ=gyro_z,
            magX=mag_x, magY=mag_y, magZ=mag_z,
        ))

    return samples


def _turn_samples(from_heading: float, to_heading: float, duration_s: float,
                  t0_ns: int = 0, noise_scale: float = 1.0) -> List[SensorData]:
    """Generate sensor data while turning in place (no steps)."""
    n = int(duration_s * SAMPLE_RATE)
    samples: List[SensorData] = []
    delta = to_heading - from_heading
    # Normalize delta
    while delta > math.pi:
        delta -= 2 * math.pi
    while delta < -math.pi:
        delta += 2 * math.pi

    yaw_rate = delta / duration_s if duration_s > 0 else 0.0

    for i in range(n):
        t_s = i / SAMPLE_RATE
        ts = t0_ns + i * DT_NS
        frac = t_s / duration_s if duration_s > 0 else 1.0
        cur_heading = from_heading + delta * frac

        # Stationary accel (gravity only, small walking to allow algorithm to NOT detect steps)
        acc_x = np.random.normal(0, ACC_NOISE * noise_scale)
        acc_y = GRAVITY + np.random.normal(0, ACC_NOISE * noise_scale * 0.3)
        acc_z = np.random.normal(0, ACC_NOISE * noise_scale)

        mag_x = MAG_STRENGTH * math.sin(cur_heading) + np.random.normal(0, MAG_NOISE * noise_scale)
        mag_y = MAG_STRENGTH * math.cos(cur_heading) + np.random.normal(0, MAG_NOISE * noise_scale)
        mag_z = np.random.normal(0, MAG_NOISE * noise_scale)

        gyro_z = yaw_rate + np.random.normal(0, GYRO_NOISE * noise_scale)

        samples.append(SensorData(
            timestamp=ts,
            accX=acc_x, accY=acc_y, accZ=acc_z,
            gyroX=0.0, gyroY=0.0, gyroZ=gyro_z,
            magX=mag_x, magY=mag_y, magZ=mag_z,
        ))
    return samples


def _heading_rad(deg: float) -> float:
    return math.radians(deg)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def simulate_straight(distance: float, heading: float, step_length: float = 0.7,
                      noise_scale: float = 1.0) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    """Returns (sensor_stream, ground_truth_points)."""
    n_steps = distance / step_length
    duration_s = n_steps / WALK_FREQ  # each step = 0.5s at 2Hz
    samples = _walk_samples(duration_s, heading, noise_scale=noise_scale)

    # Ground truth: start(0,0) -> end
    gt = [(0.0, 0.0),
          (distance * math.sin(heading), distance * math.cos(heading))]
    return samples, gt


def simulate_turn(from_heading: float, to_heading: float, duration_s: float = 1.5,
                  t0_ns: int = 0, noise_scale: float = 1.0) -> List[SensorData]:
    return _turn_samples(from_heading, to_heading, duration_s, t0_ns, noise_scale)


def simulate_l_shape(leg1: float, leg2: float, noise_scale: float = 1.0
                     ) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    heading1 = 0.0  # north
    heading2 = math.pi / 2  # east (left turn = +90° CW)

    step_length = 0.7
    dur1 = (leg1 / step_length) / WALK_FREQ
    dur2 = (leg2 / step_length) / WALK_FREQ
    turn_dur = 1.5

    s1 = _walk_samples(dur1, heading1, t0_ns=0, noise_scale=noise_scale)
    t0 = len(s1) * DT_NS
    s_turn = _turn_samples(heading1, heading2, turn_dur, t0_ns=t0, noise_scale=noise_scale)
    t0 += len(s_turn) * DT_NS
    s2 = _walk_samples(dur2, heading2, t0_ns=t0, noise_scale=noise_scale)

    samples = s1 + s_turn + s2
    gt = [(0.0, 0.0),
          (0.0, leg1),
          (leg2, leg1)]
    return samples, gt


def simulate_rectangle(width: float, height: float, noise_scale: float = 1.0
                       ) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    """Walk a rectangle: north -> east -> south -> west -> back to origin."""
    step_length = 0.7
    headings = [0.0, math.pi / 2, math.pi, 3 * math.pi / 2]
    legs = [height, width, height, width]
    turn_dur = 1.5

    samples: List[SensorData] = []
    t0 = 0

    for i, (h, leg) in enumerate(zip(headings, legs)):
        if i > 0:
            prev_h = headings[i - 1]
            st = _turn_samples(prev_h, h, turn_dur, t0_ns=t0, noise_scale=noise_scale)
            samples.extend(st)
            t0 += len(st) * DT_NS

        dur = (leg / step_length) / WALK_FREQ
        sw = _walk_samples(dur, h, t0_ns=t0, noise_scale=noise_scale)
        samples.extend(sw)
        t0 += len(sw) * DT_NS

    gt = [(0, 0), (0, height), (width, height), (width, 0), (0, 0)]
    return samples, gt


def simulate_stationary(duration_s: float, noise_scale: float = 1.0
                        ) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    n = int(duration_s * SAMPLE_RATE)
    samples: List[SensorData] = []
    for i in range(n):
        ts = i * DT_NS
        samples.append(SensorData(
            timestamp=ts,
            accX=np.random.normal(0, ACC_NOISE * noise_scale * 0.5),
            accY=GRAVITY + np.random.normal(0, ACC_NOISE * noise_scale * 0.2),
            accZ=np.random.normal(0, ACC_NOISE * noise_scale * 0.5),
            gyroZ=np.random.normal(0, GYRO_NOISE * noise_scale),
            magX=MAG_STRENGTH * 0.0 + np.random.normal(0, MAG_NOISE * noise_scale),
            magY=MAG_STRENGTH * 1.0 + np.random.normal(0, MAG_NOISE * noise_scale),
        ))
    gt = [(0, 0)]
    return samples, gt


def simulate_parking_scenario(noise_scale: float = 1.0
                              ) -> Tuple[List[SensorData], List[Tuple[float, float]]]:
    """30m north -> left turn -> 15m west -> left turn -> 10m south."""
    step_length = 0.7
    segments = [
        (0.0, 30.0),            # north 30m
        (3 * math.pi / 2, 15.0),  # west (heading=270°) 15m
        (math.pi, 10.0),        # south 10m
    ]
    turn_dur = 1.5
    samples: List[SensorData] = []
    t0 = 0

    prev_h = None
    for h, dist in segments:
        if prev_h is not None:
            st = _turn_samples(prev_h, h, turn_dur, t0_ns=t0, noise_scale=noise_scale)
            samples.extend(st)
            t0 += len(st) * DT_NS
        dur = (dist / step_length) / WALK_FREQ
        sw = _walk_samples(dur, h, t0_ns=t0, noise_scale=noise_scale)
        samples.extend(sw)
        t0 += len(sw) * DT_NS
        prev_h = h

    gt = [(0, 0), (0, 30), (-15, 30), (-15, 20)]
    return samples, gt
